t/integration/test_canvas.py

```diff
@@ -1561,12 +1561,12 @@ def test_chord_on_error(self, manager):
         assert len([cr for cr in chord_results if cr[2] != states.SUCCESS]
                    ) == 1
 
-    @pytest.mark.xfail(reason="chord header generators tasks aren't processed incrementally")
+    @flaky
     def test_generator(self, manager):
         def assert_generator(file_name):
-            for i in range(3):
+            for i in range(5):
                 sleep(1)
-                if i == 2:
+                if i == 4:
```

Comment:
https://github.com/celery/celery/issues/3021 has a very specific implementation. The first three elements are fetched before the execution of fourth. Fifth and fifth are yield together.
The first two fetches is happening in `_get_app` in lines
```
if tasks:
    app = tasks[0]._app
```
The third is happening in `header.tasks` in tasks, and then in the main loop inside _apply_tasks two tasks are fetched together 
```
_apply_tasks
```

Reply:
fixed